consensus/integration_test_types.go

```diff
@@ -0,0 +1,443 @@
+	node.lock.Lock()
+	defer node.lock.Unlock()
+
+	if node.eventLoop.status != eventLoopStatusRunning {
+		return
+	}
+
+	node.eventLoop.ProcessValidatorTimeout(timeout)
+}
+
+func (node *validatorNode) Start() {
+	node.eventLoop.Start()
+	go node.runEventSignalLoop()
+}
+
+func (node *validatorNode) runEventSignalLoop() {
+	for {
+		select {
+		case event := <-node.eventLoop.Events:
+			switch event.EventType {
+			case FastHotStuffEventTypeVote:
+				node.handleVoteEvent(event)
+				break
+			case FastHotStuffEventTypeTimeout:
+				node.handleTimeoutEvent(event)
+				break
+			case FastHotStuffEventTypeConstructVoteQC:
+				node.handleVoteQCConstructionEvent(event)
+				break
+			case FastHotStuffEventTypeConstructTimeoutQC:
+				node.handleTimeoutQCConstructionEvent(event)
+				break
+			}
+		case <-node.quit:
+			return
+		}
+	}
+}
+
+func (node *validatorNode) handleVoteEvent(event *FastHotStuffEvent) {
+	payload := GetVoteSignaturePayload(event.View, event.TipBlockHash)
+	signature, err := node.privateKey.Sign(payload[:])
+	if err != nil {
+		panic(err)
+	}
+
+	vote := &voteMessage{
+		view:      event.View,
+		blockHash: event.TipBlockHash,
+		publicKey: node.privateKey.PublicKey(),
+		signature: signature,
+	}
+
+	for _, validator := range node.validatorNodes {
+		go validator.ProcessVote(vote)
+	}
+}
+
+func (node *validatorNode) handleTimeoutEvent(event *FastHotStuffEvent) {
+	node.lock.Lock()
+	defer node.lock.Unlock()
+
+	// Skip if the timed out view from the event is stale
+	if node.eventLoop.currentView != event.View {
+		return
+	}
+
+	// Skip if the node can't advance to the next view for some reason.
+	// This should never happen.
+	if _, err := node.eventLoop.AdvanceViewOnTimeout(); err == nil {
+		node.broadcastTimeout(event)
+	}
+}
+
+func (node *validatorNode) broadcastTimeout(event *FastHotStuffEvent) {
+	payload := GetTimeoutSignaturePayload(event.View, event.QC.GetView())
+	signature, err := node.privateKey.Sign(payload[:])
+	if err != nil {
+		panic(err)
+	}
+
+	timeout := &timeoutMessage{
+		view:      event.View,
+		highQC:    event.QC,
+		publicKey: node.privateKey.PublicKey(),
+		signature: signature,
+	}
+
+	// Broadcast the block to all validators.
+	for _, validator := range node.validatorNodes {
+		go validator.ProcessTimeout(timeout)
+	}
+}
+
+func (node *validatorNode) handleVoteQCConstructionEvent(event *FastHotStuffEvent) {
+	if !node.isBlockProposer {
+		return
+	}
+
+	// Skip if the view from the event is stale. This can happen if the node
+	// has advances to the next view before the block is processed.
+	if node.eventLoop.currentView != event.View {
+		return
+	}
+
+	block := &block{
+		view:      event.View,
+		blockHash: createDummyBlockHash(),
+		height:    event.TipBlockHeight + 1,
+		qc:        event.QC,
+	}
+
+	// Broadcast the block to all validators.
+	for _, validator := range node.validatorNodes {
+		go validator.ProcessBlock(block)
```

Comment:
I know this is just a test, but if you call ProcessBlcok in a goroutine for the proposer node, then I think you can have a race condition where the following events occur:
1) onTimeoutScheduledTaskExecuted triggers, sent to server.go but not processed yet
2) ProcessValidatorVote is called, which constructs QC and sends to server.go
3) server.go processes QC via handleVoteQCConstructionEvent
4) server.go processes timeout after QC with handleTimeoutEvent

Reply:
To avoid the above race condition, you need to make sure that handleVoteQCConstructionEvent increments the view before it returns. Otherwise, a subsequent call to something like handleTimeoutEvent will erroneously conclude that the event is fresh when it's actually stale.

Reply:
That race condition isn't possible. All handlers hold the event loop's mutex when operating on it. When handling a message or an event, all handlers also check if the view on the message/event is state. So, in your example, what would happen is:
1. onTimeoutScheduledTaskExecuted triggers, sent to server.go but not processed yet
2. ProcessValidatorVote is called, which constructs QC and sends to server.go
3. server.go processes QC via handleVoteQCConstructionEvent
4. server.go sees that the timeout has a stale view (has the same view as the QC), and doesn't process the timeout